<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=gb2312">
<link href="css/report.css" rel="stylesheet" type="text/css">
<script language="javascript" src="js/jquery.js"></script>
<script language="javascript" src="js/report.js"></script>
<title>LRTK report for multiple related samples </title>
<style type="text/css">
body{
	position:relative;
	overflow:auto;
}
.note {color: #f00;}

</style>
</head>
<body>
<div class=seq_cn1>
<div class="top_btn1"></div>
<!-- Part I Overview-->
<div class="h0"><a name="overview"></a>Part I Overview</div>
<hr class="clear-contentunit" />
<h1><a name="overview.bio1"></a> Standard Bioinformatics Analysis</h1>
<p>Linked-Read ToolKit (LRTK), is a unified and versatile toolkit to analyze both metagenome and human genome linked-read sequencing data derived from any of the three major linked-read sequencing platforms. LRTK delivers a suite of utilities to perform data simulation, format conversion, data preprocessing, read cloud assembly, barcode-aware alignment, reconstruction of long fragments, and variant detection and phasing. LRTK automatically produces HTML reports to summarize the quality statistics as part of its pipeline, and generates publication-ready visualizations.</p><br/>
<br/>
<div class="tc"><a href="./img/workflow.png" target="_blank"><img style="width:720px" src="./img/workflow.png" /></a></div>
<br/>
<hr class="clear-contentunit-thin"/>
<br/>
<!--Part 2 bioinformatics analysis-->

<!--sequencing QC-->
<div class="h0"><a name="result"></a>Part II Bioinformatics Analysis Results </div>
<hr class="clear-contentunit" /> 
<h1><a name="result.abundance"></a>2.1 Taxonomic abundance</h1>
<p>We developed a reference genome-based framework to quantify the abundance of species. In brief, we align decontaminated metagenome reads to the reference genome using a barcode aware alignment pattern. We adopt a tiered alignment strategy to detect species. In the initial alignment, we map reads to the complete reference genome. We then divided the reference genome into 1 k bp windows and calculated the number of mapped reads for each window. The mapped reads are classified into two types: unique mapping reads (U) and multiple mapping reads (M). For each window W, we calculate the number of unique RC(U) and multiple RC(M) reads, respectively. For each multiple mapping read, we calculate the weight Ï‰ in terms of the number of aligned locations. To ensure proper estimation of genome abundance, we calculate the abundance as the mean of the n% most closely covered windows. After initial classification of microbial genomes, we extract the corresponding genome sequence and align the reads to the new subset reference genome sequence to refine the alignment. The final relative abundance is calculated again by using the same  processing approach.</p>
<hr class="clear-contentunit-thin"/>
<div class="break"></div>
<!-- Content unit - One column -->
Data Statistics<br/><table>
<tr><th >ID</th><th>Minimum value</th><th>Lower Quartile</th><th>Median</th><th>Upper Quartile</th><th>Maximum value</th></tr>
<tr><td>SRR14763292</td><td>7.16825241626886e-08</td><td>1.73876228990025e-05</td><td>0.00010876100170093</td><td>0.000575155933279082</td><td>0.221487146496594</td></tr><tr><td>SRR14763295</td><td>6.00229921421637e-08</td><td>9.43700347386261e-06</td><td>7.72370647953768e-05</td><td>0.000679776234062941</td><td>0.0865019276433003</td></tr><tr><td>SRR14763289</td><td>1.55780189158639e-07</td><td>1.63297539042757e-05</td><td>9.17371946498209e-05</td><td>0.000876689414510523</td><td>0.098477048457547</td></tr><tr><td>SRR14763284</td><td>5.99290152263194e-07</td><td>3.40189501154689e-05</td><td>0.000185659520384498</td><td>0.0010671191286413</td><td>0.117686832994092</td></tr><tr><td>SRR14763277</td><td>3.69610950550659e-09</td><td>1.84672497737394e-05</td><td>0.000121318460072011</td><td>0.00115059507640609</td><td>0.110546375292612</td></tr><tr><td>SRR14763290</td><td>2.21794877048371e-08</td><td>5.83451501649382e-06</td><td>6.33046446762213e-05</td><td>0.000761476826197084</td><td>0.110612455710043</td></tr><tr><td>SRR14763287</td><td>5.9953984776317e-08</td><td>1.62112103326532e-05</td><td>0.000117024515915714</td><td>0.00139945873252379</td><td>0.0631302940679857</td></tr><tr><td>SRR14763294</td><td>5.85902124966797e-08</td><td>8.91017113061345e-06</td><td>7.38879951345268e-05</td><td>0.000795891300780326</td><td>0.0469696889016275</td></tr><tr><td>SRR14763283</td><td>1.60790468028515e-08</td><td>1.03072038158038e-05</td><td>6.73809574990437e-05</td><td>0.000684772298064646</td><td>0.0608651766886825</td></tr><tr><td>SRR14763288</td><td>3.31431909986093e-08</td><td>8.49732975657646e-06</td><td>6.03714496104059e-05</td><td>0.00072680806916498</td><td>0.0893728030022723</td></tr><tr><td>SRR14763281</td><td>2.58309815215318e-08</td><td>2.03866511352412e-05</td><td>0.000131723647933235</td><td>0.00107140219312629</td><td>0.0803521613000057</td></tr><tr><td>SRR14763279</td><td>1.8998816544714e-08</td><td>1.5080576105437e-05</td><td>9.48425112720813e-05</td><td>0.000603643841378818</td><td>0.107962059506886</td></tr><tr><td>SRR14763282</td><td>7.47595874377937e-08</td><td>9.03705402399492e-06</td><td>6.29967828276221e-05</td><td>0.000706654366974488</td><td>0.080980553193661</td></tr><tr><td>SRR14763286</td><td>4.16651805117638e-07</td><td>1.65585797886086e-05</td><td>6.85483928028301e-05</td><td>0.000762132216228185</td><td>0.111206479385739</td></tr><tr><td>SRR14763291</td><td>3.89641592518545e-07</td><td>1.53709828288198e-05</td><td>7.28020833033992e-05</td><td>0.000517402237964007</td><td>0.172126964667462</td></tr><tr><td>SRR14763293</td><td>8.57533686456813e-08</td><td>7.38080157517488e-06</td><td>5.58037867277277e-05</td><td>0.000745401725194202</td><td>0.091811197488869</td></tr></table><hr class="clear-contentunit-thin" /><br/>Data distribution<br/>
<div class="tc"><a href="./img/abundance.png" target="_blank"><img style="width:720px" src="./img/abundance.png" /></a></div>
<hr class="clear-contentunit-thin" />
<h1><a name="result.abundance"></a> 2.1.1 Dimensionality Reduction </h1>
<br/>Data distribution<br/>
<div class="tc"><a href="./img/pca.png" target="_blank"><img style="width:720px" src="./img/pca.png" /></a></div>
<hr class="clear-contentunit-thin" />
<div class="break"></div>
<h1><a name="result.SNVs"></a> 2.2 Single Nucleotide Variations</h1>
LRTK supports several popular variant detection tools and promotes best practice for linked-read sequencing. For SNV and INDEL calling, LRTK provides FreeBayes, SAMtools and GATK.
<hr class="clear-contentunit-thin" />
Data Statistics<br/>
<table>
<tr><th>ID</th><th>SNV number</th><th>Homozygous rate</th><th>Heterozygous rate</th></tr>
<tr><td>SRR14763292</td><td>7.16825241626886e-08</td><td>1.73876228990025e-05</td><td>0.00010876100170093</td></tr><tr><td>SRR14763295</td><td>6.00229921421637e-08</td><td>9.43700347386261e-06</td><td>7.72370647953768e-05</td></tr><tr><td>SRR14763289</td><td>1.55780189158639e-07</td><td>1.63297539042757e-05</td><td>9.17371946498209e-05</td></tr><tr><td>SRR14763284</td><td>5.99290152263194e-07</td><td>3.40189501154689e-05</td><td>0.000185659520384498</td></tr><tr><td>SRR14763277</td><td>3.69610950550659e-09</td><td>1.84672497737394e-05</td><td>0.000121318460072011</td></tr><tr><td>SRR14763290</td><td>2.21794877048371e-08</td><td>5.83451501649382e-06</td><td>6.33046446762213e-05</td></tr><tr><td>SRR14763287</td><td>5.9953984776317e-08</td><td>1.62112103326532e-05</td><td>0.000117024515915714</td></tr><tr><td>SRR14763294</td><td>5.85902124966797e-08</td><td>8.91017113061345e-06</td><td>7.38879951345268e-05</td></tr><tr><td>SRR14763283</td><td>1.60790468028515e-08</td><td>1.03072038158038e-05</td><td>6.73809574990437e-05</td></tr><tr><td>SRR14763288</td><td>3.31431909986093e-08</td><td>8.49732975657646e-06</td><td>6.03714496104059e-05</td></tr><tr><td>SRR14763281</td><td>2.58309815215318e-08</td><td>2.03866511352412e-05</td><td>0.000131723647933235</td></tr><tr><td>SRR14763279</td><td>1.8998816544714e-08</td><td>1.5080576105437e-05</td><td>9.48425112720813e-05</td></tr><tr><td>SRR14763282</td><td>7.47595874377937e-08</td><td>9.03705402399492e-06</td><td>6.29967828276221e-05</td></tr><tr><td>SRR14763286</td><td>4.16651805117638e-07</td><td>1.65585797886086e-05</td><td>6.85483928028301e-05</td></tr><tr><td>SRR14763291</td><td>3.89641592518545e-07</td><td>1.53709828288198e-05</td><td>7.28020833033992e-05</td></tr><tr><td>SRR14763293</td><td>8.57533686456813e-08</td><td>7.38080157517488e-06</td><td>5.58037867277277e-05</td></tr></table><hr class="clear-contentunit-thin"/>
<br/>Data distribution<br/>
<div class="tc"><a href="./img/snv.png" target="_blank"><img style="width:720px" src="./img/snv.png" /></a></div>
<hr class="clear-contentunit-thin"/>
<h1><a name="result.SNVs"></a> 2.2.1 Unsupervised Clustering </h1>
<br/>Data distribution<br/>
<div class="tc"><a href="./img/clustering.png" target="_blank"><img style="width:720px" src="./img/clustering.png"/></a></div>
<h1><a name="result.SNVs"></a> 2.2.2 Mirrored allelic imbalance</h1>
<br/>Data distribution<br/>
<div class="tc"><a href="./img/MAI.png" target="_blank"><img style="width:720px" src="./img/MAI.png" /></a></div></div>
</body>
</html>
